// app/Transition.js
var Identifiable = require("app/Identifiable");
var State = require("app/State");


/**
 * A Transition for a State, on a given input to another State.
 * @constructor
 * @param {State} fromState
 * @param {State} toState
 * @param {string} min 
 * @param {string} [max] 
 */
function Transition(fromState, toState, min, max){
    // super call
    Identifiable.apply(this);

    /**
     * @type {string} min
     */
    this.min = min;

    /**
     * @type {string}
     */
    this.max = max || min;

    /**
     * The state that this transition leads to.
     * @type {State}
     */
    this.toState = toState;

    /**
     * The state that this transition comes from.
     * @type {State}
     */
    this.fromState = fromState;
}

Transition.prototype = Object.create(Identifiable.prototype);


/**
 * @param {string} input - The input to test
 * @returns {bool} true if the input applies to this Transition 
 */
Transition.prototype.accepts = function(input){
    return this.min <= input && input <= this.max;
};


/**
 * @inheritDoc
 */
Transition.prototype.toJSON = function(input){
    return JSON.stringify({
        id: this.id,
        min: this.min,
        max: this.max,
        toState: this.toState.id,
        fromState: this.fromState.id
    });
};

/**
 * @returns {string}
 */
Transition.prototype.toString = function(){
    return this.fromState + "(" + this.min + ":" + this.max + ") -> " + this.toState;
};

module.exports = Transition;
